src/clipboard.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use Clipboard;
+use SelectionData;
+use TargetEntry;
+use ffi;
+use glib::object::IsA;
+use glib::translate::*;
+use glib_ffi::gpointer;
+use libc::c_uint;
+use std::boxed::Box as Box_;
+use std::mem::transmute;
+
+pub trait ClipboardExtManual {
+    fn set_with_data<F: Fn(&Clipboard, &SelectionData, u32) + 'static>(&self, targets: &[TargetEntry], f: F) -> bool;
+}
+
+impl<O: IsA<Clipboard>> ClipboardExtManual for O {
+    fn set_with_data<F: Fn(&Clipboard, &SelectionData, u32) + 'static>(&self, targets: &[TargetEntry], f: F) -> bool {
```

Comment:
This seems more like a `set`. The with data part is an implementation detail of how the bindings implement it. No data is provided by the caller, the bindings use the data for passing a closure

Reply:
Essentially it configures a callback that provides clipboard data when
requested to do so at some point in the future. 

Regarding the name, maybe set_data_provider, set_provider, or set like you have
suggested. Though, it seems to me that either way you will have to describe its
behaviour in terms of gtk_clipboard_set_with_data, so changing the name doesn't 
seem to be of much help. I would rather remain with current one.

Reply:
I guess `set_data` is fine, or `set_with_closure`